src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceRunner.java

```diff
@@ -0,0 +1,286 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.geospatial.ip2geo.jobscheduler;
+
+import java.net.URL;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.IndicesOptions;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.geospatial.ip2geo.common.DatasourceHelper;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.common.GeoIpDataHelper;
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoSettings;
+import org.opensearch.jobscheduler.spi.JobExecutionContext;
+import org.opensearch.jobscheduler.spi.ScheduledJobParameter;
+import org.opensearch.jobscheduler.spi.ScheduledJobRunner;
+import org.opensearch.jobscheduler.spi.utils.LockService;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * Datasource update task
+ *
+ * This is a background task which is responsible for updating Ip2Geo datasource
+ */
+@Log4j2
+public class DatasourceRunner implements ScheduledJobRunner {
+
+    private static DatasourceRunner INSTANCE;
+
+    /**
+     * Return a singleton job runner instance
+     * @return job runner
+     */
+    public static DatasourceRunner getJobRunnerInstance() {
+        if (INSTANCE != null) {
+            return INSTANCE;
+        }
+        synchronized (DatasourceRunner.class) {
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
+            INSTANCE = new DatasourceRunner();
+            return INSTANCE;
+        }
+    }
```

Comment:
Let's move out of this singleton behavior and see if we can use DI. This way of creating instances reduces the ease of doing unit tests.

I would go with an old constructor way of creating the object and then use CreateComponents function in plugin class init the object only once. I understand this is common way of creating objects in Plugin and various OpenSearch plugin but we are moving away from this.

Reply:
It cannot be done such way because of a way how job-scheduler is loading extensions. Job scheduler create an instance of extension directly which prevent us from utilizing DI.

Reply:
Let me put the question like this, if I replace this code with a constructor will that create a problem?

Reply:
It won't work. You need to return instance of this class in `getJobRunner` method of DatasourceExtension class.
This DatasourceExtension class is created inside job scheduler framework which we don't have a control of.